clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```diff
@@ -182,6 +215,222 @@ public CompletableFuture<Map<TopicPartition, OffsetAndTimestampInternal>> fetchO
+            if (error != null && updatePositionsExpired) {
+                cachedUpdatePositionsException.set(error);
+            }
+        });
+    }
+
+    /**
+     * If there are partitions still needing a position and a reset policy is defined, request
+     * reset using the default policy.
+     *
+     * @param result Future that will complete when the reset operation completes.
+     * @throws NoOffsetForPartitionException If no reset strategy is configured
+     */
+    private void initWithPartitionOffsetsIfNeeded(CompletableFuture<Void> result) {
+        try {
+            // Mark partitions that need reset, using the configured reset strategy. If no
+            // strategy is defined, this will raise a NoOffsetForPartitionException exception.
+            subscriptionState.resetInitializingPositions();
+        } catch (Exception e) {
+            result.completeExceptionally(e);
+            return;
+        }
+
+        // For partitions awaiting reset, generate a ListOffset request to retrieve the partition
+        // offsets according to the strategy (ex. earliest, latest), and update the positions.
+        resetPositionsIfNeeded().whenComplete((resetResult, error) -> {
+            if (error == null) {
+                result.complete(null);
+            } else {
+                result.completeExceptionally(error);
+            }
+        });
+    }
+
+    /**
+     * Fetch the committed offsets for partitions that require initialization. Use them to set
+     * the fetch positions in the subscription state.
+     *
+     * @throws TimeoutException If offsets could not be retrieved within the timeout
+     */
+    private CompletableFuture<Void> initWithCommittedOffsetsIfNeeded(long deadlineMs) {
+        final Set<TopicPartition> initializingPartitions = subscriptionState.initializingPartitions();
+
+        if (initializingPartitions.isEmpty()) {
+            return CompletableFuture.completedFuture(null);
+        }
+
+        log.debug("Refreshing committed offsets for partitions {}", initializingPartitions);
+        CompletableFuture<Void> result = new CompletableFuture<>();
+
+        // The shorter the timeout provided to poll(), the more likely the offsets fetch will time out. To handle
+        // this case, on the first attempt to fetch the committed offsets, a FetchCommittedOffsetsEvent is created
+        // (with potentially a longer timeout) and stored. The event is used for the first attempt, but in the
+        // case it times out, subsequent attempts will also use the event in order to wait for the results.
+        CompletableFuture<Map<TopicPartition, OffsetAndMetadata>> fetchCommittedFuture;
+        if (!canReusePendingOffsetFetchEvent(initializingPartitions)) {
+            // Need to generate a new request to fetch committed offsets
+            final long fetchCommittedDeadlineMs = Math.max(deadlineMs, time.milliseconds() + defaultApiTimeoutMs);
```

Comment:
the trick here is that we want to extend the deadline (to the default_api_timeout, or the specific api call, whatever is larger). 

This pending request is mainly to solve the case of repeated calls with low timeout (ie. poll(0)), so we expect that we'll end up having the default_api_timeout used for deadline, but also considering the cases where the specific call had a larger timeout and we would use it in that case (so that's why the max). Makes sense?